core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -471,6 +473,13 @@ private void cleanup() {
         }
     }
 
+    // We directly run callbacks for event loops if we're already on the loop, which applies to the vast
+    // majority of cases.
+    private static boolean shouldRunInExecutor(@Nullable Executor executor) {
+        return executor != null &&
+               (!(executor instanceof SingleThreadEventLoop) || !((EventLoop) executor).inEventLoop());
```

Comment:
The reason we always used `Executor.execute()` was to avoid the event ordering/overlap problem due to reentrance, e.g. a `Subscriber` calls `Subscription.request()` at `Subscriber.onNext()`. Could we use a thread-local variable to make sure we are not re-entering?

Reply:
Can you describe the issue a bit more? My impression was that since events are all pushed through the queue, the ordering would be controlled by the queue, not executor, even when reentrant.

Reply:
Sure. Even if all events are pushed into a queue, `Subscriber.onNext()` can overlap:

1. `DefaultStreamMessage` calls `Subscriber.onNext()`
2. `Subscriber.onNext()` calls `Subscription.request()`
3. `Subscription.request()` may trigger `DefaultStreamMessage.notifySubscriber()`
4. `DefaultStreamMessage.notifySubscriber()` calls `Subscriber.onNext()` <-- Reentrance

Reply:
Is this a problem? Since each round through, notify will pull from the queue, it seems like order is preserved and it won't deadlock since the recursion ends after the queue is empty.

Reply:
Consider this `onNext` implementation:

```java
public void onNext(Object e) {
    subscription.request(1);
    ... Do something with 'e' ...
}
```

The later element can be handled earlier due to recursion.

Reply:
Thanks - I get it, sorry for taking a while to understand.

Do we have any determined semantics for the case where there are multi threaded writes on a stream with no executor? If not (it seems like a very bad usage of the stream), I think we can use a normal field instead of theeadlocal since we would only need to worry about single-thread event loop.

Reply:
I have added a guard to prevent reentrance - for now went with just a field, it should handle most cases except where multiple threads write simultaneously with no executor - in that case, only one thread will be used for notification whereas without the guard, multiple threads would; not clear to me whether this is a terrible behavior or not.

Reply:
Your fix won't work because of a race condition:

- Thread A sets `notifying` to `1` and enters the notification loop.
- Thread A finishes the notification loop, but it did not set `notifying` to `0` yet.
- Thread B adds some elements to the queue and calls `notifySubscriber()`, but it does nothing because `notifying` is `1` yet.
- Thread A sets `notifying` back to `0`, but it did not consume the elements produced by thread B.

Reply:
I moved the guard so it's only in the synchronous case, and also changed it from volatile to normal boolean since I found that subscription methods can only be called by a single thread at a time

https://github.com/reactive-streams/reactive-streams-jvm#2-subscriber-code  - 2.7

Also, the spec does seem to indicate that there needs to be a recursion limit, which was only true when we had the executor, so solving the non-executor case with this or similar seems required. 

https://github.com/reactive-streams/reactive-streams-jvm#3-subscription-code 3.3

But getting strange errors with the compliance tests...